opentracing/src/main/java/opentracing/SpanPropagator.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Encodes or Decodes a {@linkplain Span} in binary or text formats.
+ *
+ * <p>The toXXX methods are expected to serialize spans into a pair of values representing
+ * separately the span / span identity, and the trace attributes. This is done specifically
+ * for binary protocols that may represent tracing identity in a dedicated fixed-length slot in the
+ * binary message format, so that it can be inspected efficiently by the middleware / routing layers
+ * without parsing the whole message.
+ */
+public interface SpanPropagator {
+
+  /**
+   * Encodes the span into a binary representation of the span's identity and trace
+   * attributes.
+   */
+  byte[] toBinary(Span span);
```

Comment:
There are two reasons to split core "tracer state" from the "trace attributes":

1. allowing OT clients (whether they be application owners or maintainers of IPC frameworks) to opt-out of the "trace attribute" wire cost, regardless of the OT implementation, and
2. enabling the sorts of introspection that @yurishkuro has described many times now (no need to reiterate it here)

@michaelsembwever you say:

> I'm aware of the notion of the separation and its benefits.

... and you go on to say:

> My understanding is that the encoding specifics and any routing layer is not an end-user.

This is the crux of the problem. You don't think of RPC framework authors as users of OpenTracing, yet they are. I don't know how to convince you of this since it seems self-evident to me: they must use the OT APIs to get bits over the wire, and it is a non-starter to try and add custom methods to `SpanPropagator` for every popular IPC framework and/or middleware layer.

I think you are asserting (or at least implying) that the only "user" that matters is the application programmer who doesn't write RPC systems. If we pursue that path, I don't know how OpenTracing will be able to provide tracing interop for at the RPC/IPC layer, at least when trace attributes need to be treated separately.